From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Herman van Hazendonk <github.com@herrie.org>
Date: Tue, 20 Oct 2026 16:22:40 +0200
Subject: [PATCH] Create a channel's thread table before using it

DiscordChannel->threads is not created for every channel, but
discord_add_thread, the GUILD_CREATE thread loop, the thread list
command and the thread lookup all use it unconditionally. With a NULL
table GLib logs "hash_table != NULL" criticals and the thread is lost.

Create the table on first use in each of those places.

Signed-off-by: Herman van Hazendonk <github.com@herrie.org>
Upstream-Status: Pending
---
 libdiscord.c | 18 ++++++++++++++++--
 1 file changed, 16 insertions(+), 2 deletions(-)

diff --git a/libdiscord.c b/libdiscord.c
index 3bbfe4f..d054353 100644
--- a/libdiscord.c
+++ b/libdiscord.c
@@ -832,6 +832,9 @@ discord_add_thread(DiscordAccount *da, DiscordGuild *guild, DiscordChannel *pare
 		parent = discord_get_channel_global(da, from_int(thread->parent_id));
 	}
 	if (parent) {
+		if (parent->threads == NULL) {
+			parent->threads = g_hash_table_new_full(g_int64_hash, g_int64_equal, NULL, NULL);
+		}
 		g_hash_table_replace_int64(parent->threads, thread->id, thread);
 	}
 	return thread;
@@ -4696,8 +4699,13 @@ discord_process_dispatch(DiscordAccount *da, const gchar *type, JsonObject *data
 			guint64 parent_id = to_int(json_object_get_string_member(thread, "parent_id"));
 			DiscordChannel *parent = discord_get_channel_global_int(da, parent_id);
 
-			if (parent && !g_hash_table_contains_int64(parent->threads, thread_id)) {
-				discord_add_thread(da, guild, parent, thread, guild_id);
+			if (parent) {
+				if (parent->threads == NULL) {
+					parent->threads = g_hash_table_new_full(g_int64_hash, g_int64_equal, NULL, NULL);
+				}
+				if (!g_hash_table_contains_int64(parent->threads, thread_id)) {
+					discord_add_thread(da, guild, parent, thread, guild_id);
+				}
 			}
 		}
 
@@ -7509,6 +7517,9 @@ discord_chat_threads(PurpleConnection *pc, int id, const gchar *filter)
 	GHashTableIter thread_iter;
 	gpointer key, value;
 	/* TODO: get archived thread data */
+	if (channel->threads == NULL) {
+		channel->threads = g_hash_table_new_full(g_int64_hash, g_int64_equal, NULL, NULL);
+	}
 	g_hash_table_iter_init(&thread_iter, channel->threads);
 
 	gchar *threads_list = g_strdup(_("Active Threads:\n<pre>Creation Time       | Last Message Time   | Name"));
@@ -10049,6 +10060,9 @@ discord_get_thread_id_from_timestamp(DiscordAccount *da, PurpleConversation *con
 	GHashTableIter iter;
 	gpointer key, value;
 
+	if (channel->threads == NULL) {
+		channel->threads = g_hash_table_new_full(g_int64_hash, g_int64_equal, NULL, NULL);
+	}
 	g_hash_table_iter_init(&iter, channel->threads);
 
 	while (g_hash_table_iter_next(&iter, &key, &value)) {
-- 
2.53.0

